Separate healthz server from metrics server in kube-proxy

MrHohn opened the pull request:
From #14661, proposal is on kubernetes/community#552.

Couple bullet points as in commit:
- /healthz will be served on 0.0.0.0:10256 by default.
- /metrics and /proxyMode will be served on port 10249 as before.
- Healthz handler will verify timestamp in iptables mode.

/assign @nicksardo @bowei @thockin 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44968)
<!-- Reviewable:end -->

MrHohn commented:
Not sure what happened there:
```
I0426 14:35:22.444] Congratulations!  All Go source files have been linted.
I0426 14:35:22.444] 
I0426 14:35:22.448] FAILED   hack/make-rules/../../hack/verify-golint.sh	39s
```
@k8s-bot verify test this

bowei commented on cmd/kube-proxy/app/server.go:
```diff
@@ -359,17 +370,22 @@ func (s *ProxyServer) Run() error {
 		configz.InstallHandler(http.DefaultServeMux)
 		go wait.Until(func() {
-			err := http.ListenAndServe(s.Config.HealthzBindAddress+":"+strconv.Itoa(int(s.Config.HealthzPort)), nil)
+			err := http.ListenAndServe(s.Config.MetricsBindAddress+":"+strconv.Itoa(int(s.Config.MetricsPort)), nil)
```
Probably have the BindAddress be a string that includes "addr:port". All of the Go libraries consume this and the having it as a string gives people the option to bind to a Unix domain socket, etc

MrHohn replied:
Sounds reasonable, though in that way we have to deprecate an existing flag and change the behavior of another. Let me try a bit on this.

MrHohn replied:
Quick update: Other folks are also making this "addr:port" ([#34727/discussion](https://github.com/kubernetes/kubernetes/pull/34727#discussion_r113769514)), I will update here to match.

MrHohn replied:
Changed format to  "addr:port".

bowei commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -63,10 +63,16 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
+		obj.HealthzBindAddress = "0.0.0.0"
+	}
+	if obj.MetricsPort == 0 {
+		obj.MetricsPort = 10249
```
why 10249? why not 10257?

MrHohn replied:
10249 is what it used to be, I think it would be better to keep it (some e2e tests rely on it)?

bowei commented on pkg/proxy/healthcheck/healthcheck.go:
```diff
@@ -233,3 +236,100 @@ func (hcs *server) SyncEndpoints(newEndpoints map[types.NamespacedName]int) erro
+// updated, it verifies we don't exceed max no respond duration since
+// last update.
+type HealthzServer struct {
+	listener    Listener
```
do you need locking?

MrHohn replied:
The plan is to use sync/atomic.Value for concurrent read/write. I think it should be sufficient?

thockin commented:
Approving on spec - this is not a code review :)

Make sure flags are backwards-compatible (though it seems you are not changing the flag?)

thockin commented:
/approve

MrHohn commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1495,6 +1498,11 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
 	proxier.portsMap = replacementPortsMap
 
+	// Update healthz timestamp if it is periodic sync.
+	if proxier.healthzServer != nil && reason == syncReasonForce {
```
@nicksardo Moved `UpdateTimestamp` to the end of `syncProxyRules` so that it wouldn't be updated if anything fails in between, thanks!

MrHohn commented:
> Make sure flags are backwards-compatible (though it seems you are not changing the flag?)

Most of the flags are marked as deprecated by #34727. All relevant parameters could now be configured through a config file. Hence no flags are added.

MrHohn commented:
All tests passed :)
PR is ready for review.

nicksardo commented on pkg/proxy/healthcheck/healthcheck.go:
```diff
@@ -233,3 +236,100 @@ func (hcs *server) SyncEndpoints(newEndpoints map[types.NamespacedName]int) erro
+			"lastUpdated": %q,
+			"currentTime": %q
+		}
+		`, lastUpdated, currentTime)), "\n"))
```
Is dedenting necessary? I think any parser can handle whitespace. This HTTP handler could potentially be called a lot, so response writing should be as cheap as possible.

MrHohn replied:
Makes sense, modified response to a single line.

bowei commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -64,9 +64,14 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
 		obj.BindAddress = "0.0.0.0"
 	}
 	if len(obj.HealthzBindAddress) == 0 {
```
seems more clear to just compare against empty string `obj.HealthzBindAddress == ""`

MrHohn replied:
Done.

bowei commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -64,9 +64,14 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
-		obj.HealthzBindAddress = ":10249"
+		obj.HealthzBindAddress = ":10256"
+	}
+	if len(obj.MetricsBindAddress) == 0 {
```
see above

MrHohn replied:
Done.

bowei commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -64,9 +64,14 @@ func SetDefaults_KubeProxyConfiguration(obj *KubeProxyConfiguration) {
+	if len(obj.MetricsBindAddress) == 0 {
+		obj.MetricsBindAddress = "127.0.0.1:10249"
+	} else if !strings.Contains(obj.MetricsBindAddress, ":") {
+		obj.MetricsBindAddress = ":10249"
```
don't you mean `+=` ?

MrHohn replied:
Thanks, fixed this part.

bowei commented on pkg/master/ports/ports.go:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
-	// ProxyPort is the default port for the proxy healthz server.
+	// ProxyStatusPort is the default port for the proxy metrics server.
 	// May be overridden by a flag at startup.
 	ProxyStatusPort = 10249
```
why don't we use these constants to generate the strings in the defaults above?

MrHohn replied:
Done.

bowei commented on pkg/proxy/healthcheck/healthcheck.go:
```diff
@@ -233,3 +236,95 @@ func (hcs *server) SyncEndpoints(newEndpoints map[types.NamespacedName]int) erro
+}
+
+// NewDefaultHealthzServer returns a default healthz http server.
+func NewDefaultHealthzServer(addr string, maxNoRespondDuration time.Duration) *HealthzServer {
```
maxNoRespondDuration => healthTimeout

seems to be more descriptive

MrHohn replied:
Sounds good, changed to healthTimeout.

bowei commented on pkg/proxy/healthcheck/healthcheck.go:
```diff
@@ -233,3 +236,95 @@ func (hcs *server) SyncEndpoints(newEndpoints map[types.NamespacedName]int) erro
+	var lastUpdated time.Time
+	val := h.hs.lastUpdated.Load()
+	if val == nil {
+		lastUpdated = time.Time{}
```
```
lastUpdated := time.Time{}
if val := h.hs.lastUpdated.Load(); val != nil {
  lastUpdated = val.(time.Time)
}
```

MrHohn replied:
Thanks, this looks much better. Fix applied.

bowei commented:
/lgtm

MrHohn committed:
Add healthz server support to healthcheck package

MrHohn committed:
Separate healthz server from metrics server in kube-proxy

- /healthz will be served on 0.0.0.0:10256 by default.
- /metrics and /proxyMode will be served on port 10249
  as before.
- Healthz handler will verify timestamp in iptables mode.

MrHohn committed:
Update kube-proxy healthz port in e2e test

MrHohn committed:
Autogenerated files

MrHohn commented:
@bowei Thanks. Squashed review commits.

MrHohn commented:
@k8s-bot gce etcd3 e2e test this

MrHohn commented:
@k8s-bot gce etcd3 e2e test this

MrHohn commented:
@bowei Could you re-lgtm? Thanks

bowei commented:
/lgtm

jimmycuadra commented:
This change broke our production deployment of Kubernetes: https://github.com/kubernetes/kubernetes/issues/48600#issuecomment-318253103
